pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/ASTVariableDeclaratorId.java

```diff
@@ -104,13 +121,29 @@ public ASTType getTypeNode() {
         return null;
     }
 
-    private Node findTypeNameNode(Node node) {
-        int i = 0;
-        while (node.jjtGetChild(i) instanceof ASTAnnotation) {
-            // skip annotations
-            i++;
-        }
-        ASTType typeNode = (ASTType) node.jjtGetChild(i);
-        return typeNode.jjtGetChild(0);
+    // @formatter:off
+    /**
+     * Returns the type of this node. The type of a declarator ID is
+     * <ul>
+     *   <li>1. not necessarily the same as the type written out at the
+     *          start of the declaration, e.g. {@code int a[];}
+     *   <li>2. not necessarily the same as the types of other variables
+     *          declared in the same statement, e.g. {@code int a[], b;}.
+     * </ul>
+     *
+     * <p>These are consequences of Java's allowing programmers to
+     * declare additional pairs of brackets on declarator ids. The type
+     * of the node returned by {@link #getTypeNode()} doesn't take into
+     * account those additional array dimensions, whereas this node's
+     * type takes into account the total number of dimensions, i.e.
+     * those declared on this node plus those declared on the type node.
+     *
+     * <p>The type of the declarator ID is thus always the real type of
+     * the variable.
+     */
+    // @formatter:on
+    @Override
+    public Class<?> getType() {
+        return super.getType();
```

Comment:
on this case however, the doc is actually useful as the relation between code / type may be non-obvious on these valid, yet corner cases. Kudos on the detail